libs/s25main/ingameWindows/iwMapGenerator.cpp

```diff
@@ -52,11 +53,12 @@ enum
     CTRL_RIVERS,
     CTRL_MOUNTAIN_DIST,
     CTRL_TREES,
-    CTRL_STONE_PILES
+    CTRL_STONE_PILES,
+    CTRL_ISLANDS
 };
 
 iwMapGenerator::iwMapGenerator(MapSettings& settings)
-    : IngameWindow(CGI_MAP_GENERATOR, IngameWindow::posLastOrCenter, Extent(270, 470), _("Map Generator"),
+    : IngameWindow(CGI_MAP_GENERATOR, IngameWindow::posLastOrCenter, Extent(270, 520), _("Map Generator"),
```

Comment:
Is there a nice way to test the `IngameWindow` sub classes (like simulating change of combo selections)? I looked into `Test_UI/testIngameWindow` but didn't manage to call any members due to access modifiers (private/protected).

Reply:
What exactly do you want to do? You can use `GetCtrl` or `GetCtrls` to get a specific one and change that and/or call `Msg_ButtonClick` and friends directly. They are public (at least in `Window`)

Reply:
When I do something like this:
```
uiHelper::initGUITests();
rttr::mapGenerator::MapSettings settings;
iwMapGenerator wnd(settings);    
auto* combo = wnd.GetCtrl<ctrlComboBox>(CTRL_ISLANDS);
combo->SetSelection(2);
```

I get an error for the last line of code saying:
`Member access into incomplete type 'ctrlComboBox'`

Reply:
"incomplete type" means you are missing the include. --> `#include "controls/ctrlComboBox.h"` in the test will solve this

Reply:
Still having a problem with the `BOOST_TEST`. Somehow it doesn't compile (not even on CI server). `BOOST_REQUIRE` seems to work, that's why I used it in the past I believe.

Reply:
I guess you already found my comment at https://github.com/Return-To-The-Roots/s25client/pull/1369#issuecomment-779374470 ?
The difference between those 2 macros is that the latter will print the values on failure. I.e. you can write `BOOST_TEST(foo == 2)` and it will tell you the value of `foo` when the assertion fails, which is very valuable when checking a failure. It is equivalent to `BOOST_CHECK_EQUAL(foo, 2)` which does the same. `BOOST_REQUIRE` does only check, if the assertion is true or false and doesn't tell you anything about the values. This makes it hard, when a test fails.

For that the types must be printable, that's what the compile error says. Especially strong enums/enum classes are not printable by default, hence you need the custom output op or exclude those from being printed. This was the reason I added the GameTypesOutput header

Reply:
Yeah got that from your previous comment! Was too lazy to override the `<<` operator for those 3 types, but I should probably do that.